### PR TITLE
Support for encoding directives, arg groups, length-prefixed e-expressions

### PR DESCRIPTION
**Pull request by zslayton**
* Adds support for switching encodings midstream
* Adds support for parameter cardinality modifiers
* Adds support for reading argument groups
* Adds support for length-prefixed e-expressions
* Optimizes and simplifies macro evaluation
* Simplifies struct expansion logic
* Adds static analysis optimizations to template evaluation
* Adds support for lazily evaluating top-level e-expressions
* Adds more Ion 1.1 encodings to the `read_many_structs` benchmark
* Shrinks many data types

### CODE REVIEW

**Comment by zslayton** [comment]
Miri is failing due to #794. I'd like to investigate that after this PR is merged.

File: src/lazy/expanded/mod.rs
Comment (zslayton) [inline src/lazy/expanded/mod.rs:785]: 🗺️ `is_null`, `has_annotations`, `annotations`, and `ion_type` can all be implemented without needing to actually evaluate the singleton macro. We remember these values from the compilation phase.

File: src/lazy/expanded/template.rs
Comment (zslayton) [inline src/lazy/expanded/template.rs:306]: 🗺️ And again.

**Review by zslayton — COMMENTED** [review]
🗺️ PR Tour 🧭

File: src/lazy/expanded/compiler.rs
Comment (popematt) [inline src/lazy/expanded/compiler.rs:248]: This is TDL, right?
```suggestion
        // TODO: Syntactic support for address IDs like `(14 ...)`
```

File: src/lazy/expanded/macro_evaluator.rs
Comment (popematt) [inline src/lazy/expanded/macro_evaluator.rs:1124]: Can you change the formatting so that the indentation is consistent?

File: src/lazy/expanded/macro_evaluator.rs
Comment (popematt) [inline src/lazy/expanded/macro_evaluator.rs:1171]: Can you make this name more descriptive and/or add a comment?

File: src/lazy/any_encoding.rs
Comment (zslayton) [reply src/lazy/any_encoding.rs:429]: It doesn't, but the reason is actually sort of interesting (for the right audience 😅). The reader knows to change encodings whenever it encounters an IVM _in the current encoding_. `E0 01 00 EA` and `E0 01 01 EA` are syntactically valid in both binary Ion 1.0 and 1.1. `$ion_1_0` and `$ion_1_1` are syntactically valid in both text 1.0 and text 1.1. However, if the binary reader encountered `$ion_1_0` it wouldn't recognize it and would surface an error. As a result, no additional logic was needed to prevent switching between text and binary.

File: src/lazy/binary/raw/v1_1/immutable_buffer.rs
Comment (zslayton) [reply src/lazy/binary/raw/v1_1/immutable_buffer.rs:472]: The `ImmutableBuffer` and `TextBufferView` types both need a reference to the bump allocator in order to:
* (text) Decode escape sequences
* (text) Decode base64
* (text) Join multi-section long-form strings into one value.
* (any) Bump-allocate large data types so we can pass around a `&'top` reference instead.
* (any) Bump-allocate macro-constructed values from `(make_string)`, `(make_list)`, `(annotate)`, etc.

Starting with this PR, the buffers need to be able to parse macro invocations too. That means that all parsing methods would need to accept a reference to the macro table in order to successfully parse a value which could have a macro invocation at any level of nesting. (Including: `read`, `read_sequence_value`, `peek_sequence_value`, `read_struct`, `read_struct_field`, etc etc.) Importantly, data types that would lazily call a parsing method (`LazyValue`, `LazyList`, `LazyStruct`, .... `LazyExpandedValue`, `LazyExpandedList` .... `LazyRawValue`, `LazyRawList`, etc) would also need to hold/track a reference to the macro table for the moment they eventually decided to call a parsing method.

Rather than selectively plumb through distinct `allocator` and `macro_table` references throughout, I chose to simply give each buffer a reference to the encoding context. By definition, the encoding context is 'frozen' throughout the codebase when parsing occurs, so there's not much downside to liberally handing out references to it.

File: src/lazy/binary/raw/v1_1/value.rs
Comment (zslayton) [reply src/lazy/binary/raw/v1_1/value.rs:189]: > I don't think it's too long to put inline in read_resolved

Sorry, what's not too long?

`read_resolved_general_case` handles _all_ possibilities (including string and int). This is only creating a more-likely-to-be-inlined path for strings and ints.

File: src/lazy/decoder.rs
Comment (zslayton) [reply src/lazy/decoder.rs:75]: I don't think this is as clear-cut as that.

> Personally, I would get rid of the new_ prefix so that only "old" functions are prefixed, but I don't have a strong preference how you do it so long as it's consistent.

If someone calls `ivm.encoding()`, I think they could reasonably expect that to mean the encoding that was used to encode the IVM. However, under this proposal it actually means the encoding that the _next_ stream segment will use.

The methods which do not have a prefix are:
* `is_text`
* `is_binary`
* `major_minor`

`is_text` and `is_binary` have no concept of "new"/"old" -- they MUST agree for the stream segment leading up to the IVM and the stream segment that follows it.

`major_minor` is the major-minor version being marked by the IVM. I'm also having a hard time finding a way to misinterpret this one--if there were new/old versions of it, they would also be forced to always agree.

I could get behind changing the affected names to refer to the stream segments, i.e.:
* `stream_encoding_before`/`stream_encoding_after`
* `stream_version_before`/`stream_version_after`
* `is_text`/`is_binary`
* `major_minor`

What do you think of that?

File: src/lazy/any_encoding.rs
Comment (zslayton) [reply src/lazy/any_encoding.rs:535]: This is re-exported at the root, but I agree it should live in its own module and/or at the root. I'd like to leave that for another PR though.

File: src/lazy/any_encoding.rs
Comment (zslayton) [reply src/lazy/any_encoding.rs:602]: There's a doc comment on the trait method (which shows up on hover in RustRover, fwiw)

File: src/lazy/binary/raw/v1_1/e_expression.rs
Comment (zslayton) [reply src/lazy/binary/raw/v1_1/e_expression.rs:49]: Hm, this comment explains the inner `cache` field which is not `pub`. I've added a short doc comment for the type itself, however.

File: src/lazy/decoder.rs
Comment (popematt) [reply src/lazy/decoder.rs:75]: > major_minor is the major-minor version being marked by the IVM. I'm also having a hard time finding a way to misinterpret this one--if there were new/old versions of it, they would also be forced to always agree.

I think I was interpreting this as being the `major_minor` of the Ion encoding (which one—before or after?) rather than of the IVM and that's why I was getting confused. 

I think that adding `_before` and `_after` is a very helpful clarification (and it's easier to catch the implied "right now" that exists between "before" and "after").

File: src/lazy/binary/raw/v1_1/immutable_buffer.rs
Comment (zslayton) [reply src/lazy/binary/raw/v1_1/immutable_buffer.rs:61]: `context` is basically a collection of resources; an allocator, a reference to the macro table, etc. An argument could be made that two buffers are not equal if they're holding references to different contexts, but this is a very low-level feature-gated construct so I'm ok with leaving it as arguably imperfect.

File: src/lazy/binary/raw/v1_1/immutable_buffer.rs
Comment (zslayton) [reply src/lazy/binary/raw/v1_1/immutable_buffer.rs:294]: Oops, good catch.

File: src/lazy/binary/raw/v1_1/type_code.rs
Comment (zslayton) [reply src/lazy/binary/raw/v1_1/type_code.rs:38]: Yeah, it should be further down. There was a hot second when I think this opcode was `0xEE`, which is how it ended up here.

File: src/lazy/binary/raw/v1_1/value.rs
Comment (popematt) [reply src/lazy/binary/raw/v1_1/value.rs:189]: > Sorry, what's not too long?

Well...

> `read_resolved_general_case` is defined in an inherent impl block further below. In retrospect, I could have defined it nested inside `read_resolved`. What do you think?

I don't think the definition of `read_resolved_general_case` is too long to define nested inside `read_resolved`—i.e. it won't hurt the readability of the code to do so. I also think that (assuming there are no terrible performance implications) it would be fine to put all of the branches right here rather than having them defined in a separate function.

File: src/lazy/encoder/write_as_ion.rs
Comment (zslayton) [reply src/lazy/encoder/write_as_ion.rs:344]: Agreed, I'll do this with the other moves in a future PR

File: src/lazy/binary/raw/v1_1/immutable_buffer.rs
Comment (popematt) [reply src/lazy/binary/raw/v1_1/immutable_buffer.rs:61]:
```suggestion
        // A definition of equality that ignores the `context` field.
        // An argument could be made that two buffers are not equal if they're holding references to
        // different contexts, but this is a very low-level, feature-gated construct so it's probably
        // fine if the implementation is arguably imperfect.
```

File: src/types/symbol.rs
Comment (zslayton) [reply src/types/symbol.rs:16]: Comment updated:
```rust
// TODO: Turn this into a Box<str>.
//       Symbols are read-only, so there's no chance we'll add data to the `String`. 
//       Using a `Box<str>` shrinks this value from 24 bytes to 8 bytes.
```

File: src/lazy/expanded/e_expression.rs
Comment (zslayton) [reply src/lazy/expanded/e_expression.rs:40]: No; in order to return a fully resolved reference to the macro, it needs to get the `&'top Macro` reference to the macro living in the macro table.

File: src/lazy/expanded/mod.rs
Comment (zslayton) [reply src/lazy/expanded/mod.rs:406]: >  Then you could interpret them in the match block around L502:

This isn't an option because `next_value` is a convenience method that most applications will not exercise. (Everything goes through `next_item`, but only some tooling and other low-level stuff uses the `next_value` wrapper.

File: src/lazy/expanded/mod.rs
Comment (zslayton) [reply src/lazy/expanded/mod.rs:507]: Per our offline discussion, this (and some of the weirdness that spurred the introduction of custom macros) came about because `?` was showing up in some profiler hotspots. I'd like to leave them in place for the sake of merging this PR, but am open to trying to re-introduce `?` in another PR while looking out for performance regressions.

File: src/lazy/expanded/mod.rs
Comment (zslayton) [reply src/lazy/expanded/mod.rs:541]: Same issue here with `?`.

File: src/lazy/expanded/mod.rs
Comment (zslayton) [reply src/lazy/expanded/mod.rs:573]: Same issue here with `?`.

File: src/lazy/expanded/mod.rs
Comment (zslayton) [reply src/lazy/expanded/mod.rs:599]: Same issue here with `?`.

File: src/lazy/text/buffer.rs
Comment (zslayton) [reply src/lazy/text/buffer.rs:1076]: Removed, thanks.

File: src/lazy/expanded/template.rs
Comment (zslayton) [reply src/lazy/expanded/template.rs:168]: Good idea; I wrote this before the other cardinalities were supported. Added.

File: src/lazy/any_encoding.rs
Comment (popematt) [reply src/lazy/any_encoding.rs:343]: As per offline discussion, this is a performance issue rather than an ergonomics issue.

File: src/lazy/expanded/macro_evaluator.rs
Comment (zslayton) [reply src/lazy/expanded/macro_evaluator.rs:1171]: I have no idea what that name referred to. x_x

I'm pretty sure it was a one-off to debug something. I've removed it.

File: src/lazy/expanded/template.rs
Comment (zslayton) [reply src/lazy/expanded/template.rs:28]: Yes -- `ParameterCardinality` is `!`/`?`/`*`/`+`, while `RestSyntaxPolicy` indicates whether the value is both non-`!` _and_ appears in a tail position.

You could derive the `RestSyntaxPolicy` by looking at `ParameterCardinality` and the `&Parameter`'s position in the larger `MacroSignature`, but in many places (for size savings) types only hold the `&Parameter`. Storing this as its own field simplifies parsing.

**Review by popematt — APPROVED** [review]
Approving so that we can review PR feedback changes in smaller PRs.

**Pull request merged**
